lib/src/repo.rs

```diff
@@ -1032,6 +1032,40 @@ impl MutableRepo {
         new_ids
     }
 
+    /// Fully resolves transitive replacements in `parent_mapping`.
+    ///
+    /// If `parent_mapping` contains cycles, this function will panic.
+    fn resolve_rewrite_mapping_with(
+        &self,
+        mut predicate: impl FnMut(&Rewrite) -> bool,
+    ) -> HashMap<CommitId, Vec<CommitId>> {
+        let sorted_ids = dag_walk::topo_order_forward(
+            self.parent_mapping.keys(),
+            |&id| id,
+            |&id| match self.parent_mapping.get(id).filter(|&v| predicate(v)) {
+                None => &[],
+                Some(rewrite) => rewrite.new_parent_ids(),
+            },
+        );
+        let mut new_mapping: HashMap<CommitId, Vec<CommitId>> = HashMap::new();
+        for old_id in sorted_ids {
+            let Some(rewrite) = self.parent_mapping.get(old_id).filter(|&v| predicate(v)) else {
+                continue;
+            };
+            let lookup = |id| new_mapping.get(id).map_or(slice::from_ref(id), |ids| ids);
```

Comment:
Does `unwrap_or()` work here?

Reply:
No. We need to coerce `&Vec<_>` to `&[_]`.